xds/src/main/java/io/grpc/xds/internal/rbac/engine/AuthorizationEngine.java

```diff
@@ -0,0 +1,218 @@
+public class AuthorizationEngine<ReqT, RespT> {
+  private static final Logger log = Logger.getLogger(AuthorizationEngine.class.getName());
+  
+  /**
+   * RbacEngine is an inner class that holds RBAC action 
+   * and conditions of RBAC policy.
+   */
+  private static class RbacEngine {
+    @SuppressWarnings("UnusedVariable")
+    private final Action action;
+    private final ImmutableMap<String, Expr> conditions;
+
+    public RbacEngine(Action action, ImmutableMap<String, Expr> conditions) {
+      this.action = action;
+      this.conditions = conditions;
+    }
+  }
+
+  private final RbacEngine allowEngine;
+  private final RbacEngine denyEngine;
+
+  /**
+   * Creates a CEL-based Authorization Engine from a list of Envoy RBACs.
+   * The constructor can take either one or two Envoy RBACs.
+   * When it takes two RBACs, the order has to be a RBAC with DENY action
+   * followed by a RBAC with ALLOW action.
+   * @param rbacPolicies input Envoy RBAC list.
+   * @throws IllegalArgumentException if the user inputs an invalid RBAC list.
+   */
+  public AuthorizationEngine(ImmutableList<RBAC> rbacPolicies) throws IllegalArgumentException {
+    if (rbacPolicies.size() < 1 || rbacPolicies.size() > 2) {
+      throw new IllegalArgumentException(
+        "Invalid RBAC list size, must provide either one RBAC or two RBACs. ");
+    } 
+    if (rbacPolicies.size() == 2 && (rbacPolicies.get(0).getAction() != Action.DENY 
+        || rbacPolicies.get(1).getAction() != Action.ALLOW)) {
+      throw new IllegalArgumentException( "Invalid RBAC list, " 
+          + "must provide a RBAC with DENY action followed by a RBAC with ALLOW action. ");
+    }
+    RbacEngine allowEngine = null;
+    RbacEngine denyEngine = null;
+    for (RBAC rbac : rbacPolicies) {
+      Map<String, Expr> conditions = new LinkedHashMap<>();
+      for (Map.Entry<String, Policy> rbacPolicy: rbac.getPolicies().entrySet()) {
+        conditions.put(rbacPolicy.getKey(), rbacPolicy.getValue().getCondition());
+      }
+      if (rbac.getAction() == Action.ALLOW) {
+        allowEngine = new RbacEngine(Action.ALLOW, ImmutableMap.copyOf(conditions));
+      }
+      if (rbac.getAction() == Action.DENY) {
+        denyEngine = new RbacEngine(Action.DENY, ImmutableMap.copyOf(conditions));
+      }
+    }
+    this.allowEngine = allowEngine;
+    this.denyEngine = denyEngine;
+  }
+
+  /**
+   * The evaluate function performs the core authorization mechanism of CEL Evaluation Engine.
+   * Determines whether a gRPC call is allowed, denied, or unable to decide.
+   * @param args evaluate argument that is used to evaluate the RBAC conditions.
+   * @return an AuthorizationDecision generated by CEL Evaluation Engine.
+   */
+  public AuthorizationDecision evaluate(EvaluateArgs<ReqT, RespT> args) {
+    List<String> unknownPolicyNames = new ArrayList<>();
+    // Set up activation used in CEL library's eval function.
+    Activation activation = Activation.copyOf(extractFields(args));
```

Comment:
Not in this PR but `from` might be a better name than `copyOf` since the method creates an `Activation` from a `Map`?

Reply:
Yes, I agree it makes more sense to call it `from` rather than `copyOf`. However, this function is originally from the CEL library, so I'm not able to change it unfortunately :(